modules/core/test/test_io.cpp

```diff
@@ -799,6 +801,25 @@ TEST(Core_InputOutput, filestorage_base64_basic_memory_JSON)
     test_filestorage_basic(cv::FileStorage::WRITE_BASE64, ".json", true, true);
 }
 
+// issue #21851
+TEST(Core_InputOutput, filestorage_heap_overflow)
+{
+    const ::testing::TestInfo* const test_info = ::testing::UnitTest::GetInstance()->current_test_info();
+    CV_Assert(test_info);
+
+    std::string name = std::string(test_info->test_case_name()) + "--" + test_info->name();
+    const char data[] = {0x00, 0x2f, 0x4a, 0x4a, 0x50, 0x4a, 0x4a };
+
+    std::ofstream file;
+    file.open(name, std::ios_base::binary);
+    assert(file.is_open());
+
+    file.write(data, sizeof(data));
+    file.close();
+
+    // This just shouldn't segfault, otherwise it's fine
+    EXPECT_ANY_THROW(FileStorage(name, FileStorage::READ));
```

Comment:
`FileStorage` could be initialized from memory buffer. Is the issue is reproducible with bytes array too? If yes, then you can do the test in memory and do not write to disc. See `FileStorage::MEMORY` option: https://docs.opencv.org/4.x/da/d56/classcv_1_1FileStorage.html#a973e41cb75ef6230412a567723b7482d and  example: https://github.com/opencv/opencv/blob/66f3c2673cd424a9c282fb0c8622a4ae43d3e81d/samples/cpp/filestorage.cpp#L165

Reply:
This code gets executed only if we're opening the file. Otherwise we return early from this block:
https://github.com/opencv/opencv/blob/66f3c2673cd424a9c282fb0c8622a4ae43d3e81d/modules/core/src/persistence.cpp#L784